### ISSUE TITLE
LangChain

### ISSUE THREAD

**Issue by YunchaoYang** [YunchaoYang/Blogs#51]
# What is langchain
LangChain is a framework for developing applications powered by language models. 
This framework consists of LangChain Libraries, [LangChain Templates](https://github.com/langchain-ai/langchain/blob/master/templates), [LangServe](https://github.com/langchain-ai/langserve), and [LangSmith](https://smith.langchain.com/).

1. Does LangChain Implement RAG algorithms? Yes
2. LangChain provides standard, extendable interfaces and integrations for the following modules: [Model I/O](https://python.langchain.com/docs/modules/model_io/), [Retrieval](https://python.langchain.com/docs/modules/data_connection/),[Agents](https://python.langchain.com/docs/modules/agents/).

**Comment by YunchaoYang** [comment]
# SentenceTransformers 

## What is SentenceTransformers 
SentenceTransformers is a Python library that provides pre-trained models for generating dense vector representations of sentences or text. These dense vector representations, also known as embeddings, capture the semantic meaning of the input text. These embeddings can then be used for various natural language processing (NLP) tasks such as semantic similarity calculation, clustering, classification, and information retrieval.

## Why need SentenceTransformers ?
SentenceTransformers is needed because it simplifies the process of generating high-quality sentence embeddings. Before the advent of SentenceTransformers and similar libraries, generating effective embeddings for sentences required significant expertise in NLP, deep learning, and access to large-scale compute resources for training models from scratch. SentenceTransformers provides pre-trained models that have been trained on large text corpora, which saves time and resources for practitioners and researchers. By using pre-trained models, users can obtain meaningful sentence representations without the need for extensive training data or computational resources.

Furthermore, SentenceTransformers offers a user-friendly interface that allows developers to easily integrate sentence embedding functionality into their NLP applications, making it accessible to a wider range of users. Overall, SentenceTransformers simplifies the process of leveraging advanced NLP techniques for tasks that require understanding the semantic similarity or meaning of sentences.

## What are the models used in  SentenceTransformers 
SentenceTransformers primarily uses transformer-based architectures, such as BERT (Bidirectional Encoder Representations from Transformers) and RoBERTa (Robustly optimized BERT approach), for generating embeddings of sentences or text. These transformer models are pre-trained on large corpora of text data using unsupervised learning techniques, such as masked language modeling and next sentence prediction.

## Usage
### Install
```
pip install -U sentence-transformers
```
### use it 

``` python
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("all-MiniLM-L6-v2")

# Our sentences to encode
sentences = [
    "This framework generates embeddings for each input sentence",
    "Sentences are passed as a list of string.",
    "The quick brown fox jumps over the lazy dog."
]

# Sentences are encoded by calling model.encode()
embeddings = model.encode(sentences)

# Print the embeddings
for sentence, embedding in zip(sentences, embeddings):
    print("Sentence:", sentence)
    print("Embedding:", embedding)
    print("")
```

### Pretrained Models
https://www.sbert.net/docs/pretrained_models.html

**Comment by YunchaoYang** [comment]
具体而言，该框架由以下开源库组成：

langchain-core：基本抽象和 LangChain 表达式语言。
langchain-community：第三方集成。
合作伙伴包（例如 langchain-openai，langchain-anthropic 等）：某些集成已进一步拆分为仅依赖于 langchain-core 的轻量级包。
langchain：构成应用程序认知架构的链条、代理和检索策略。
[langgraph(opens in a new tab)](https://python.langchain.com/v0.2/docs/langgraph)：通过将步骤建模为图中的边缘和节点，使用LLMs构建强大且有状态的多角色应用程序。
[langserve(opens in a new tab)](https://python.langchain.com/v0.2/docs/langserve/)：将 LangChain 链条部署为 REST API。
[LangSmith](https://www.langchain.asia/langsmith)：开发人员平台，可让你对LLM应用程序进行调试、测试、评估和监控，并与LangChain无缝集成。

LangChain 表达式语言（LCEL）是许多 LangChain 组件的基础，是一种组合链式的声明式方法。LCEL 从一开始就被设计为支持将原型投入到生产中，从最简单的“提示 + LLM”链条到最复杂的链条，无需更改代

**Comment by YunchaoYang** [comment]
![Image](https://github.com/user-attachments/assets/46d3a05b-119a-4935-96d4-e5f0ebfdf85b)

![Image](https://github.com/user-attachments/assets/32494260-4c66-4b6a-8599-e9028461aa12)
